Feature/randfft cherry pick

bjoo opened the pull request:
Cherry picked the cuFFT, cuRand and quda_fp16 refactors from my hip oriented WIP branch.
Code built on Summit, but some tests on Summit had errors... @maddyscientist said she hasn't checked it for correctness recently. I'll roll back to her version (from before my refactor) and check if that had an issue, or whether what I picked up is a Summit issue.

Commit:
FFT Refactor

Commit:
cuRandom refactor

Commit:
Cherry Picked FFT Refactor, changed to use TARGET_CUDA

Commit:
FFT, Random and FP16 refactors

Commit:
Removed switchbox headers and renamed the target files as per request from  @maddyscientist

maddyscientist commented on include/kernels/pgauge_init.cuh:
```diff
@@ -121,7 +121,11 @@ namespace quda {
+    // Was   xurand(*localState>& 1 ? 1 : -1
+    // which presumably just selects when the lowest bit is 1 or 0 with 50% probability each
+    // so this should do the same, without an appeal to the bit swizzle, but may end up being slower.
+    stheta = ( uniform<T>::rand(localState) < static_cast<T>(0.5) ? 1 : -1 ) * sqrt( (T)1.0 - ctheta * ctheta );
```
What's the motivation for this change?

maddyscientist commented on include/targets/hip/random_quda.h:
```diff
@@ -0,0 +1,152 @@
+#pragma once
```
There's a lot of repeated boilerplate in here and the `CUDA` equivalent.  While of course the `hiprandState` vs `cuRNGState` needs to be in the specialization, why not move the class `RNG` to the generic header?  We could make the class templated, with the type used to instantiate the class either `cuRNGState` or `hiprandState` in the respective target specialization?

bjoo commented:
That is possible to do, but I’d rather do it in a cleanup as I am keen to return to the other branch
I am working on.

Best,
B

--
Balint Joo,  Oak Ridge Leadership Computing Facility,
Oak Ridge National Laboratory
P.O. Box 2008, 1 Bethel Valley Road, Oak Ridge, TN 37831, USA
email: joob AT ornl.gov.  Tel: +1-757-912-0566 (cell, remote)

From: maddyscientist <notifications@github.com>
Reply-To: lattice/quda <reply@reply.github.com>
Date: Monday, October 26, 2020 at 11:12 PM
To: lattice/quda <quda@noreply.github.com>
Cc: Balint Joo <bjoo@jlab.org>, Author <author@noreply.github.com>
Subject: [EXTERNAL] Re: [lattice/quda] Feature/randfft cherry pick (#1069)
Resent-From: <bjoo@jlab.org>
Resent-Date: Monday, October 26, 2020 at 11:12 PM


@maddyscientist commented on this pull request.

________________________________

In include/targets/hip/random_quda.h<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_lattice_quda_pull_1069-23discussion-5Fr512391063&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=v4_sF8XzsuLyu3A-RZPTtzYSKw2d7I8LNg7rDgc5C0g&s=c8JgT9YFV6qEruicJPnLoSsS3NMoYcEpcd8WbDzEmd0&e=>:

> @@ -0,0 +1,152 @@

+#pragma once

There's a lot of repeated boilerplate in here and the CUDA equivalent. While of course the hiprandState vs cuRNGState needs to be in the specialization, why not move the class RNG to the generic header? We could make the class templated, with the type used to instantiate the class either cuRNGState or hiprandState in the respective target specialization?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_lattice_quda_pull_1069-23pullrequestreview-2D517318085&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=v4_sF8XzsuLyu3A-RZPTtzYSKw2d7I8LNg7rDgc5C0g&s=4G1uA-TpGRgwtPFYpsqrJkWUdavQ1e1A5xps3NVShZw&e=>, or unsubscribe<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AAEPL2HCGSAGHBH5AYGC24TSMY3BDANCNFSM4S73GAPA&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=v4_sF8XzsuLyu3A-RZPTtzYSKw2d7I8LNg7rDgc5C0g&s=hhyKfDszfL3wT8pgPeXCK5gyoN5Ey3KpQR7QEFD0aJM&e=>.

bjoo commented:
We discussed this in the meeting.
pgauge_init originally called curand() directly (which was API specific) and the
purpose appeared to be purely to make a 50-50 choice (based on the lowest bit
being 1 or 0). This change

  *   removes the API specificity by calling the API independent uniform() function
  *   makes it clear that we are trying to make a 50-50 choice by trying to “throw under” 0.5 rather than bitmasking
  *

--
Balint Joo,  Oak Ridge Leadership Computing Facility,
Oak Ridge National Laboratory
P.O. Box 2008, 1 Bethel Valley Road, Oak Ridge, TN 37831, USA
email: joob AT ornl.gov.  Tel: +1-757-912-0566 (cell, remote)

From: maddyscientist <notifications@github.com>
Reply-To: lattice/quda <reply@reply.github.com>
Date: Monday, October 26, 2020 at 11:08 PM
To: lattice/quda <quda@noreply.github.com>
Cc: Balint Joo <bjoo@jlab.org>, Author <author@noreply.github.com>
Subject: [EXTERNAL] Re: [lattice/quda] Feature/randfft cherry pick (#1069)
Resent-From: <bjoo@jlab.org>
Resent-Date: Monday, October 26, 2020 at 11:08 PM


@maddyscientist commented on this pull request.

________________________________

In include/kernels/pgauge_init.cuh<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_lattice_quda_pull_1069-23discussion-5Fr512390019&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=SYYO5g4mH1r6K7T6wRpFB0Z8lwvLdF1ri62Vc6PTlYo&s=9957IrIdJe5DH52KfJIHZxFG8y7sRKBBkbmhEmCoKPE&e=>:

> @@ -121,7 +121,11 @@ namespace quda {

     aabs = sqrt( 1.0 - a(0,0) * a(0,0));

     ctheta = Random<T>(localState, (T)-1.0, (T)1.0);

     phi = PII * Random<T>(localState);

-    stheta = ( curand(&localState) & 1 ? 1 : -1 ) * sqrt( (T)1.0 - ctheta * ctheta );

+

+    // Was   xurand(*localState>& 1 ? 1 : -1

+    // which presumably just selects when the lowest bit is 1 or 0 with 50% probability each

+    // so this should do the same, without an appeal to the bit swizzle, but may end up being slower.

+    stheta = ( uniform<T>::rand(localState) < static_cast<T>(0.5) ? 1 : -1 ) * sqrt( (T)1.0 - ctheta * ctheta );

What's the motivation for this change?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_lattice_quda_pull_1069-23pullrequestreview-2D517316924&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=SYYO5g4mH1r6K7T6wRpFB0Z8lwvLdF1ri62Vc6PTlYo&s=Xj7qfak99L4GKdIIxQ1A4FokBok3eB51ZLFWOm1l3Rg&e=>, or unsubscribe<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AAEPL2HVHXC2CWSSUTHBFOTSMY2S3ANCNFSM4S73GAPA&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=SYYO5g4mH1r6K7T6wRpFB0Z8lwvLdF1ri62Vc6PTlYo&s=0dDMvFL9Q4W_yBkHBBTLV0EFcwEPpB9qOUVak1UgM1U&e=>.

Commit:
Pulling in Kate's fixes

Merge branch 'feature/generic_kernel' into feature/randfft_cherry_pick

Commit:
Refactored common RNG class into include random_quda.h
Left curand and hiprand implementations of uniform:: and normal:: in include/targets/<arch>/random_quda_impl.h
Adapted gauge and spinor algorithms to use uniform<T>::rand() instead of Random<T>
since they were implemented identically
Removed Random<T> and left uniform<T>::rand() in place to reduce duplication/boilerplate.
Summit tests pass with the exclption of blas_test_wilson_fill (which also failed before refactor) (@maddyscientist is aware)

bjoo commented:
As discussed on Slack:
* moved RNG class back into `include/random_quda.h` since it is common between QUDA and HIP targets (may end up being completely arch independent)

* as result renamed `include/targets/cuda/random_quda.h` and `include/targets/hip/random_quda.h` to be `random_quda_impl.h` respectively in both directories

* got rid of `Random<T>(state)` and `Random<T>(state,a,b)`  since these are uniform random deviates implemented identically in `uniform<T>::rand(state)` (and now also `uniform<T>::rand(state,a,b)`. What is more with care similar functionality (to scale and translate a distribution could be implemented in `normal<T>::`. This reduced the size of the CUDA/HIP specific implementations in the `random_quda_impl.h` files.
* trawled the code and eliminated calls to `Random<T>(...)` and replace them with `uniform<T>::rand(...)`
* Code compiled on summit and everything passed except 
```The following tests FAILED:
	  2 - blas_test_full_wilson (Failed)
```
which also failed for me on summit in the regular `feature/generic_kernel` branch prior to merges. Filed this as issue #1070

Commit:
Merge branch 'feature/generic_kernel' into feature/randfft_cherry_pick

bjoo commented:
With latest merge commit, all tests passed on Summit! I hope this should be good to go now.

Commit:
Added stubs to CUFFT_Plans.h for when GAUGE_ALG is not enabled

maddyscientist committed:
Add compile time check that a QUDA_TARGET has been set
